### PR TITLE
SC-592: Add dynamic proxying support

### PR DESCRIPTION
**Pull request by hexaglow**
Derived from the original PR #255 

This PR adds the infrastructure and code to allow dynamic proxying to occur. Dynamic proxying is the ability to proxy requests for RPCs that the proxy doesn't know about in advance.

- The router package allows us to serve RPCs using using a dynamic routing table, based on service name and device name.
- The reflectionapi package allows us to change the advertised types the grpc server hosting the gateway returns from its reflection api
- The controller has new hooks to allow systems to adjust the above at runtime
- The node package has API changes to allow dynamically registering both routed and unrouted services
- The gateway system has been rewritten to allow the use of reflection to discover and proxy RPCs

Implementation Notes

The new proxy system code is organised into two stages: scan and announce.

The scan stage performs a number of RPCs against the hub and enrolled nodes to discover the names and types associated with each node. Everything it discovers is placed into the structs defined in model.go.

The announce stage watches the model and updates the node with names and metadata, and announces the discovered services and devices on the Node.

### CODE REVIEW

**Comment by hexaglow** [comment]
Made a new PR because I couldn't request review from @mattnathan as he opened the original PR

**Comment by hexaglow** [comment]
Current (major) problem is that because we deserialise request messages into an `emptypb.Empty`, all interceptors that rely on access to request fields are broken. This includes the security policies.

**Comment by hexaglow** [comment]
Fixed the interceptor problem by using `dynamicpb` instead of `emptypb` so field access by name is available.

File: pkg/system/gateway/announce.go [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+	defer r.mu.Unlock()
+	r.m[k]--
+	if r.m[k] <= 0 {
+		delete(r.m, k)
+	}
+	return r.m[k]
+}
+
+// fullNameToRpcPath maps proto full names (package.Service.Method) to rpc paths (/package.Service/Method).
+func fullNameToRpcPath(fullName protoreflect.FullName) string {
```
Comment (mattnathan) [inline pkg/system/gateway/announce.go:393]: This looks to be unused

File: proto/devices.proto [hunk 2]
```diff
@@ -21,6 +21,8 @@ service DevicesApi {
 message Device {
   string name = 1;
   smartcore.traits.Metadata metadata = 2;
+  // The name of the node that this device is announced from.
+  string node = 3;
```
Comment (mattnathan) [inline proto/devices.proto:25]: I don't think this change is related and deserves some more thought to implement properly as a "where did it come from" concept. I can't see anything in the rest of the PR that requires this, can you remove it please?

**Review by mattnathan — CHANGES_REQUESTED** [review]

File: proto/devices.proto (hunk 2 above)
Comment (hexaglow) [reply proto/devices.proto:25]: Yep, added this by mistake

File: internal/router/stream.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+			msgCap := &captureMsgRecver{MsgRecver: src}
+			client, err = resolver(msgCap)
+			if err != nil {
+				return err
+			}
+			if msg, ok := msgCap.msg.(proto.Message); ok {
+				// this case means the resolver read a message to do the resolve
+				m = msg
+			} else if msgCap.msg == nil {
+				// this case means no message was read, so we need to do it ourselves
```
[the remark this answers is not in the document]
Comment (hexaglow) [reply internal/router/stream.go:143]: Not sure I understand, the comment is already in the negative?

**Review by mattnathan — APPROVED** [review]

**Pull request merged**